### PR TITLE
Replaced static test classes with mocks

### PR DESCRIPTION
**Pull request by StefanSpieker**
I wanted to fix two **spotbugs** issues but thought that rewriting the test would be more useful than just fixing the two errors. I added some comments to clarify what the test really tests.

### Proposed changelog entries

* Internal: Reworked a test

### CODE REVIEW

**Review by oleg-nenashev — COMMENTED** [review]
Not sure about it. We generally try to minimize usage of Mockito and Powermovk in the integration tests suite

**Comment by StefanSpieker** [comment]
In general I totally agree, but the existing Test was based on a manual implemented mock class. So from my point of view I just used a library which achieved the same with a lot less lines of code. But I added the 

> CALLS_REAL_METHODS

so that in any case the real method of the abstract class will be called. I think it is even "more real" than before ;)

**Review by res0nance — APPROVED** [review]
I think mocking in this case is fine because there is a mocked class anyway.

File: core/src/test/java/hudson/slaves/NodeListTest.java [hunk 1]
```diff
@@ -24,124 +24,53 @@
...
 
         File tmp = File.createTempFile("test","test");
         try {
             XmlFile x = new XmlFile(Jenkins.XSTREAM, tmp);
             x.write(nl);
 
             String xml = FileUtils.readFileToString(tmp, Charset.defaultCharset());
-            System.out.println(xml);
-            assertEquals(6,xml.split("\n").length);
+            // System.out.println(xml); // just for debug
```
Comment (timja) [inline core/src/test/java/hudson/slaves/NodeListTest.java]: this still needed?
```suggestion
```

**Review by jvz — APPROVED** [review]

File: core/src/test/java/hudson/slaves/NodeListTest.java [hunk 2]
```diff
@@ -24,124 +24,52 @@
...
-            assertEquals(6,xml.split("\n").length);
+            // check that at least some content
+            assertTrue(xml.split("\n").length > 6);
 
             NodeList back = (NodeList)x.read();
 
+            // there should only be the 'normal' node
             assertEquals(1,back.size());
-            assertEquals(DummyNode.class,back.get(0).getClass());
+            assertEquals(dummyNode.getClass(),back.get(0).getClass());
```
Comment (alecharp) [inline core/src/test/java/hudson/slaves/NodeListTest.java]:
```suggestion
            assertEquals(dummyNode.getClass(), back.get(0).getClass());
```

File: core/src/test/java/hudson/slaves/NodeListTest.java [hunk 3]
```diff
@@ -24,124 +24,52 @@
...
         File tmp = File.createTempFile("test","test");
         try {
             XmlFile x = new XmlFile(Jenkins.XSTREAM, tmp);
             x.write(nl);
 
             String xml = FileUtils.readFileToString(tmp, Charset.defaultCharset());
-            System.out.println(xml);
-            assertEquals(6,xml.split("\n").length);
+            // check that at least some content
+            assertTrue(xml.split("\n").length > 6);
```
Comment (alecharp) [inline core/src/test/java/hudson/slaves/NodeListTest.java]: We used to test that we had exactly 6, now at least 7. I'm not sure to understand.

Comment (MRamonLeon) [inline core/src/test/java/hudson/slaves/NodeListTest.java]: why the change in the assertion?

File: core/src/test/java/hudson/slaves/NodeListTest.java [hunk 4]
```diff
@@ -24,124 +24,52 @@
...
             NodeList back = (NodeList)x.read();
 
+            // there should only be the 'normal' node
             assertEquals(1,back.size());
-            assertEquals(DummyNode.class,back.get(0).getClass());
+            assertEquals(dummyNode.getClass(),back.get(0).getClass());
         } finally {
-            tmp.delete();
+            if (!tmp.delete()) {
+                System.out.println("unable to delete File: " + tmp.getName());
```
Comment (MRamonLeon) [inline core/src/test/java/hudson/slaves/NodeListTest.java]: No need to do that I think. You also can use a     
```
@Rule
public TemporaryFolder tmp = new TemporaryFolder();
```
to do it automatically

**Review by alecharp — COMMENTED** [review]
I'm not sure that `CALLS_REAL_METHODS` on interfaces (`Node`) will have the same behavior but it will mark the test as failing is any of those methods are called. So, it should be ok.

File: core/src/test/java/hudson/slaves/NodeListTest.java (hunk 3 above)
Comment (StefanSpieker) [reply core/src/test/java/hudson/slaves/NodeListTest.java]: Because the mock xml is bigger than 6 lines and I wanted to just check that it is not smaller than 6 lines but not that precise because I want not to fix this line for every mockito update.

Comment (StefanSpieker) [reply core/src/test/java/hudson/slaves/NodeListTest.java]: Because the mock xml is bigger than 6 lines and I wanted to just check that it is not smaller than 6 lines but not that precise because I want not to fix this line for every mockito update. It is actually around 21 lines now because it is a much bigger "implementation" as before.

Comment (MRamonLeon) [reply core/src/test/java/hudson/slaves/NodeListTest.java]: makes sense

**Review by MRamonLeon — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Looks like we have a consensus here. 
I plan to merge it tomorrow if no negative feedback
